### PR TITLE
Preliminary trims as keys for lua

### PR DESCRIPTION
**Pull request by pfeerick**
Largely based on open-flysky implementation

Can probably be improved, and I don't believe some of the key events are working. 

However, the crucial ones for ELRS are, and it automagically switches the right trim up and down to up and down.

### CODE REVIEW

**Comment by JimB40** [comment]
Is there mode-switching between trims and key-emulation or joysticks works only in key-emulation mode?

**Comment by pfeerick** [comment]
Automatically switches from trims to key input "emulation" on lua standalone.

File: radio/src/targets/nv14/keys_driver.cpp [hunk 1]
```diff
@@ -25,33 +25,58 @@
...
-    result |= 1 << KEY_DOWN;
-  if (TRIMS_GPIO_REG_RVU & TRIMS_GPIO_PIN_RVU)
-    result |= 1 << KEY_UP;
-  if (TRIMS_GPIO_REG_RHL & TRIMS_GPIO_PIN_RHL)
-    result |= 1 << KEY_LEFT;
-  if (TRIMS_GPIO_REG_RHR & TRIMS_GPIO_PIN_RHR)
-    result |= 1 << KEY_RIGHT;
+  bool getKeys = true;
+#if defined(LUA)
+  if (scriptInternalData[0].reference != SCRIPT_STANDALONE) {
```
[the remark this answers is not in the document]
Comment (pfeerick) [reply radio/src/targets/nv14/keys_driver.cpp:30]: As in a 'bool luaStandloneRunning' type function?

[the remark this answers is not in the document]
Comment (raphaelcoeffic) [reply radio/src/targets/nv14/keys_driver.cpp:30]: Well, if you correct the typos and name it `luaIsStandaloneRunning()`, then yes ;-)

**Review by raphaelcoeffic — APPROVED** [review]

**Comment by JimB40** [comment]
Okay any additional info abt 'how it works?'
Left joy press - EVT_VIRTUAL_EXIT
Right joy press - EVT_VIRTUAL_ENTER
Right joy down - EVT_VIRTUAL_NEXT
Right joy up - EVT_VIRTUAL_PREV

anything else? Can't get MENU, NEXT_PAGE, PREV_PAGE or TELE

**Comment by raphaelcoeffic** [comment]
> Okay any additional info abt 'how it works?' Left joy press - EVT_VIRTUAL_EXIT Right joy press - EVT_VIRTUAL_ENTER Right joy down - EVT_VIRTUAL_NEXT Right joy up - EVT_VIRTUAL_PREV
> 
> anything else? Can't get MENU, NEXT_PAGE, PREV_PAGE or TELE

That could be mapped to the left joystick (up/down/left/right).

**Comment by pfeerick** [comment]
That's like the third time I swear I replied to something on github, and it's not there when I look the next day. 

@raphaelcoeffic Ccn you look at why I don't seem to be getting KEY_TELEM, KEY_PGDN or KEY_PGUP to play nicely with the events? I do get an event when I press the trim hats, but it doesn't get matched. VSCode has been hinting at me to keep looking at the taranis `board.h` for those defines/EnumKeys rather than the nv14 `board.h`, which is not confusing whatsover :-/

Anyway, I think I finally have a handle on the virtual events, so I just need some clarity on what will map to what... This is what I would propose at present

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Left J Up | MENU |
| Left J Down | TELEM |
| Left J Left | INC|
| Left J Right | DEC
| Left J Push | EXIT

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Right J Up | PREV |
| Right J Down | NEXT |
| Right J Left | PREV_PAGE |
| Right J Right | NEXT_PAGE |
| Right J Push | ENTER |

**Comment by raphaelcoeffic** [comment]
> Ccn you look at why I don't seem to be getting KEY_TELEM, KEY_PGDN or KEY_PGUP to play nicely with the events? I do get an event when I press the trim hats, but it doesn't get matched.

I believe the events need to declared in a specific order for that to work properly, but I can have a look.

**Comment by pfeerick** [comment]
I did about two commits back, did a full DFU flash to make sure hadn't buggered bootloader up ;)

**Comment by pfeerick** [comment]
Bootloader still works fine. Thanks to Raphael's review last night, this is now working like I wanted. Worked perfectly with ELRS v1 and ES915TX HappyModel ELRS module. 

This is how key to event mapping now stands (usually triggered on break, and with some LONG events also on long press):

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Left J Up | PAGE_UP |
| Left J Down | PAGE_DOWN  |
| Left J Left | MENU, MENU_LONG |
| Left J Right | |
| Left J Push | EXIT |

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Right J Up | PREV / DEC, possibly _RPT also |
| Right J Down | NEXT / INC, possibly _RPT also |
| Right J Left | |
| Right J Right | |
| Right J Push | ENTER, ENTER_LONG |

**Comment by JimB40** [comment]
Hey @pfeerick, 
Works as expected.
I've given quite a long test run on different apps. Anch checked "Radios Data LUA DEV" sheet to recall how it's done across other radios. Seems like it should be kind of Xlite-X12S marriage :)

So feedback & findings.

1. Right joystick should serve as 'navigate/edit'

Assignments in navigation state
Right Joy UP - Next Item
Right Joy DOWN - Previous Item
Right Joy RIGHT - Next Page
Right Joy LEFT - Previous Page
Right Joy DOWN - Select/Accept 

Assignments in edit state
Right Joy UP - higher Value
Right Joy DOWN - Lower value
Right Joy RIGHT - to be defined (may serve as MAX value)  
Right Joy LEFT - to be defined (may serve as MIX value)
Right Joy DOWN - Accept value

2. Left joystick should serve as "rest of keys" 

Left Joy UP - free, to be defined
Left Joy DOWN - TELE key
Left Joy RIGHT - MDL key 
Left Joy LEFT - SYS key
Left Joy DOWN - Cancel / EXIT key 

- EXIT on press on left joy is okay as it is separate from navigation/change
- In current version INC is down and DEC up which is counter-intuitive (same was with tango)

LUA/C++ events/constants
To be discussed if we can implement proper EVT_xxx_FIRST,EVT_xxx_LONG,EVT_xxx_REPT,EVT_xxx_BREAK event que. Then to assign EVT_VIRTUAL_xxx to some of them. It's valuable for LUA as some of EVT_VIRTUAL_xxx are not always enough to code proper UX behaviour. Haven't checked in source code if all numeric key events have EVT_xxx constants assigned, but I think only some of them are assigned. 

- Left joy key 'rest of keys' assignments would be a great addition. Then UI on NV14 would be fully navigable and we would 'buy' some time to make touchUI proper. Plus great enhancement for dev testing

- Last but not least. I was thinking how to enable OS wide joystick navigation without sacrificing trims.
There is one slot (UP) free on left joy but I think I have better idea if feasible. :)
NV14 has those power buttons. So if they are accessible we could code Left PWR press for Trim-mode & Right PWR press for Navigation-mode. Probably long press to avoid accidental change.

**Comment by pfeerick** [comment]
A lot of this is not relevant - this PR is about LUA scripts, so any mention of UI navigation outside of that, changes to event queues, or other parts of code will be ignored in this PR as it is out of scope, and should be brought up in new issues. I do not want to delay implementation of this for LUA any further, whilst we discuss other UI navigation implementation concepts. 

> I've given quite a long test run on different apps

Such as? I'm interested in knowing what apps compatible that are suitable for use for testing also ;) You haven't by any chance updated your dev tools script to suit the NV14 aspect ratio have you?

> Assignments in navigation state 
> Assignments in edit state

I'm not sure this is actually differentiated between by the system, so may not be possible. Also, it would currently be unique to NV14, no? as the underlying RIGHT and LEFT are PGUP, PGDN, and do not change to MIN/MAX on any other radio AFAIK. 

> In current version INC is down and DEC up which is counter-intuitive (same was with tango)

That can be changed, but I personally thing generally all the others are the wrong way around... up goes DOWN a list, and down goes UP a list? WTH? So no wonder the Tango was the same. 

> Are there C++ constants EVT_VIRTUAL_PAGE_UP, EVT_VIRTUAL_PAGE_DOWN? Error? You created new ones instead EVT_VIRTUAL_NEXT_PAGE, EVT_VIRTUAL_PREV_PAGE?

No, there are not, that was a typo... :grin: As you would have clearly seen since any script looking for `PREV_PAGE` and `NEXT_PAGE` would have worked ;) 

So I will move NEXT, PREV to right joystick, and flip INC/DEC then merge so it's in the next nightly. Then we can discuss the other topics in their own issues.

**Comment by pfeerick** [comment]
By 
> In current version INC is down and DEC up which is counter-intuitive (same was with tango)

Were you referring to actual behaviour, or documented, as I had those listed wrong in the table. In all radios VIRTUAL_PREV is same key as VIRTUAL_INC, and VIRTUAL_NEXT is same key as VIRTUAL_DEC, which is how they were set.

Same as with VIRTUAL_MENU - that follows MODEL key. 

This is how it stands now. 

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Left J Up | |
| Left J Down | |
| Left J Left | MENU, MENU_LONG |
| Left J Right | |
| Left J Push | EXIT |

| Stick | Maps to `EVT_VIRTUAL_` | 
| --- | --- |
| Right J Up | PREV/ INC, possibly _RPT also |
| Right J Down | NEXT/ DEC, possibly _RPT also |
| Right J Left | PREV_PAGE |
| Right J Right | NEXT_PAGE |
| Right J Push | ENTER, ENTER_LONG |

**Comment by JimB40** [comment]
You're right this general topic. Used word LUA because what I can see only what is reported as event by LUA, still I had general solution in the back of my had. :)  Not super proficient in C++ reading yet but want to be on the same page not to propose something which time-cosuming or not feasible. Discussion thread then?

Prepared after your PR alpha of TBS AL and SD buddy which are still crashing :)
I will adjust DEV Tools for NV14 as this will be valuable tool.

**Comment by pfeerick** [comment]
Now tested with Betaflight Lua also on NV14, and both it, ELRSv1 and a NV14 tweaked version of JimB40s lua devtools are working great with joystick navigation, and it automagically is just working.

**Pull request merged**